util/progress/progress.go

```diff
@@ -18,22 +18,28 @@ type contextKeyT string
 
 var contextKey = contextKeyT("buildkit/util/progress")
 
-// FromContext returns a progress writer from a context.
-func FromContext(ctx context.Context, opts ...WriterOption) (Writer, bool, context.Context) {
-	v := ctx.Value(contextKey)
-	pw, ok := v.(*progressWriter)
-	if !ok {
-		if pw, ok := v.(*MultiWriter); ok {
-			return pw, true, ctx
+// WriterFactory will generate a new progress writer.  It is the callers
+// responsibility to Close the returned Writer to avoid resource leaks.
+type WriterFactory func() (Writer, bool, context.Context)
+
+// FromContext returns a ProgressFactory to generate new progress writers
+func FromContext(ctx context.Context, opts ...WriterOption) WriterFactory {
+	return func() (Writer, bool, context.Context) {
+		v := ctx.Value(contextKey)
+		pw, ok := v.(*progressWriter)
+		if !ok {
+			if pw, ok := v.(*MultiWriter); ok {
+				return pw, true, ctx
+			}
+			return &noOpWriter{}, false, ctx
 		}
-		return &noOpWriter{}, false, ctx
-	}
-	pw = newWriter(pw)
-	for _, o := range opts {
-		o(pw)
+		pw = newWriter(pw)
+		for _, o := range opts {
+			o(pw)
+		}
+		ctx = context.WithValue(ctx, contextKey, pw)
```

Comment:
I think this should return `(WriterFactory, bool, context.Context)`. It is wrong for the callback to return context that was created by another stack. That goroutine is likely already completed and original context has been canceled. If you have a case where you need new context when `newWriter` is called then the callback itself should take a context and new writer is added to that context, not to the one that was passed into `FromContext`.

Reply:
I see that `newWriter()` result is added to the context. If that is the requirement and there isn't another way to reorganize this that means that my second proposal makes more sense. For the default case, you can add back `NewFromContext` helper so that caller does not need to pass context twice when from&new happen together.

Reply:
Okay, makes sense.   Update the WriterFactory now takes a `ctx` and added `NewFromContext`  wrapper to automatically call the WriterFactory with the same `ctx` passed in.